src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -0,0 +1,474 @@
+
+            public async Task<ImmutableArray<DiagnosticData>> GetSpecificDiagnosticsAsync(CancellationToken cancellationToken)
+            {
+                if (CurrentSolution == null)
+                {
+                    return ImmutableArray<DiagnosticData>.Empty;
+                }
+
+                var argsId = Id as LiveDiagnosticUpdateArgsId;
+                if (argsId == null)
+                {
+                    return ImmutableArray<DiagnosticData>.Empty;
+                }
+
+                if (CurrentProject == null)
+                {
+                    // when we return cached result, make sure we at least return something that exist in current solution
+                    return ImmutableArray<DiagnosticData>.Empty;
+                }
+
+                var stateSet = this.StateManager.GetOrCreateStateSet(CurrentProject, argsId.Analyzer);
+                if (stateSet == null)
+                {
+                    return ImmutableArray<DiagnosticData>.Empty;
+                }
+
+                var diagnostics = await GetDiagnosticsAsync(stateSet, CurrentProject, CurrentDocumentId, (AnalysisKind)argsId.Kind, cancellationToken).ConfigureAwait(false);
+                if (diagnostics == null)
+                {
+                    // Document or project might have been removed from the solution.
+                    return ImmutableArray<DiagnosticData>.Empty;
+                }
+
+                return FilterSuppressedDiagnostics(diagnostics.Value);
+            }
+
+            public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(CancellationToken cancellationToken)
+            {
+                if (CurrentSolution == null)
+                {
+                    return ImmutableArray<DiagnosticData>.Empty;
+                }
+
+                if (CurrentProjectId != null)
+                {
+                    if (CurrentProject == null)
+                    {
+                        return GetDiagnosticData();
+                    }
+
+                    var documentIds = CurrentDocumentId != null ? SpecializedCollections.SingletonEnumerable(CurrentDocumentId) : CurrentProject.DocumentIds;
+
+                    // return diagnostics specific to one project or document
+                    await AppendDiagnosticsAsync(CurrentProject, CurrentDocumentId, documentIds, cancellationToken).ConfigureAwait(false);
+                    return GetDiagnosticData();
+                }
+
+                await AppendDiagnosticsAsync(CurrentSolution, cancellationToken).ConfigureAwait(false);
+                return GetDiagnosticData();
+            }
+
+            protected async Task AppendDiagnosticsAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                // PERF: should run this concurrently? analyzer driver itself is already running concurrently.
```

Comment:
Analyzer driver will run concurrently only when CompilationWithAnalyzersOptions is created with ConcurrentAnalysis = true. We should ensure that any UI thread operations that explicitly ask for diagnostics (e.g. FixAll, possibly Ctrl + Dot), run with ConcurrentAnalysis = true and background thread ones with ConcurrentAnalysis = false.